File: app/src/main/java/com/battlelancer/seriesguide/jobs/episodes/SeasonBaseJob.java
package com.battlelancer.seriesguide.jobs.episodes;

import android.content.Context;
import androidx.annotation.NonNull;
import com.battlelancer.seriesguide.provider.SgRoomDatabase;
import com.battlelancer.seriesguide.provider.SgSeason2Numbers;

/**
 * Flagging whole seasons watched or collected.
 */
public abstract class SeasonBaseJob extends BaseEpisodesJob {

    protected final long seasonId;
    private SgSeason2Numbers season;

    public SeasonBaseJob(long seasonId, int flagValue, JobAction action) {
        super(flagValue, action);
        this.seasonId = seasonId;
    }

    @Override
    public boolean applyLocalChanges(Context context, boolean requiresNetworkJob) {
        SgSeason2Numbers season = SgRoomDatabase.getInstance(context).sgSeason2Helper()
                .getSeasonNumbers(seasonId);
        if (season == null) {
            return false;
        }
        this.season = season;

        return super.applyLocalChanges(context, requiresNetworkJob);
    }

    @NonNull
    public SgSeason2Numbers getSeason() {
        return season;
    }

    @Override
    protected long getShowId() {
        return getSeason().getShowId();
    }

    public long getSeasonId() {
        return seasonId;
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/jobs/episodes/SeasonBaseJob.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2017, 2018, 2021-2024 Uwe Trottmann
package com.battlelancer.seriesguide.jobs.episodes

import android.content.Context
import com.battlelancer.seriesguide.provider.SgRoomDatabase
import com.battlelancer.seriesguide.shows.database.SgSeason2Numbers

/**
 * Flagging whole seasons watched or collected.
 */
abstract class SeasonBaseJob(
    val seasonId: Long,
    flagValue: Int,
    action: JobAction
) : BaseEpisodesJob(flagValue, action) {

    private lateinit var season: SgSeason2Numbers
    override fun applyLocalChanges(context: Context, requiresNetworkJob: Boolean): Boolean {
        val season = SgRoomDatabase.getInstance(context).sgSeason2Helper()
            .getSeasonNumbers(seasonId) ?: return false
        this.season = season
        return super.applyLocalChanges(context, requiresNetworkJob)
    }

    fun getSeason(): SgSeason2Numbers {
        return season
    }

    override val showId: Long
        get() = getSeason().showId
}
